### PR TITLE
Support automated JWT login

### PR DESCRIPTION
**Pull request by LLipter**
Pipeline #1019338

### CODE REVIEW

File: python/mujinwebstackclient/controllerwebclientraw.py [hunk 1]
```diff
@@ -248,3 +249,19 @@ def CallGraphAPI(self, query, variables=None, headers=None, timeout=5.0):
             raise ControllerGraphClientException(_('Unexpected server response %d: %s') % (statusCode, raw), statusCode=statusCode, response=response)
 
         return content['data']
+
+    def Login(self, timeout=5):
```
Comment (woswos) [inline python/mujinwebstackclient/controllerwebclientraw.py:253]: @LLipter Can we implement a custom auth class derived from `AuthBase` and use it for `self._session.auth` instead?

Comment (woswos) [reply python/mujinwebstackclient/controllerwebclientraw.py:253]: The current method is going to make a login request even if no API calls are made using the client, which is wasteful

Comment (LLipter) [reply python/mujinwebstackclient/controllerwebclientraw.py:274]: @woswos Fixed. Could you have a look again?

Comment (woswos) [reply python/mujinwebstackclient/controllerwebclientraw.py:274]: @LLipter I think we can handle both JWT and basic auth login entirely in the custom auth class without the login functions to further simplify the design.

Comment (LLipter) [reply python/mujinwebstackclient/controllerwebclientraw.py]: done

File: python/mujinwebstackclient/controllerwebclientraw.py [hunk 2]
```diff
@@ -125,6 +153,10 @@ def Request(self, method, path, timeout=5, headers=None, **kwargs):
             kwargs['allow_redirects'] = method in ('GET',)
 
         response = self._session.request(method=method, url=url, timeout=timeout, headers=headers, **kwargs)
+        jsonWebToken = response.cookies.get('jwttoken')
```
Comment (woswos) [inline python/mujinwebstackclient/controllerwebclientraw.py:156]: @LLipter You can also handle this part in the custom auth class with a response hook

File: python/mujinwebstackclient/controllerwebclientraw.py [hunk 3]
```diff
@@ -26,6 +26,42 @@
...
 log = logging.getLogger(__name__)
 
+class JsonWebTokenAuth(requests_auth.AuthBase):
+    """Attaches JWT Bearer Authentication to a given Request object. Use basic authentication if token is not available."""
+
+    _username = None
+    _password = None
+    _jsonWebToken = None
+
+    def __init__(self, username, password, jsonWebToken):
```
Comment (woswos) [inline python/mujinwebstackclient/controllerwebclientraw.py:36]: Do we need the `jsonWebToken` argument?

File: python/mujinwebstackclient/controllerwebclientraw.py [hunk 4]
```diff
@@ -26,6 +26,42 @@
 import logging
 log = logging.getLogger(__name__)
 
+class JsonWebTokenAuth(requests_auth.AuthBase):
+    """Attaches JWT Bearer Authentication to a given Request object. Use basic authentication if token is not available."""
+
+    _username = None
```
Comment (woswos) [inline python/mujinwebstackclient/controllerwebclientraw.py:32]: Please add comments  for the class members

File: python/mujinwebstackclient/controllerwebclientraw.py [hunk 5]
```diff
@@ -26,6 +26,42 @@
...
+    def __ne__(self, other):
+        return not self == other
+
+    def __call__(self, request):
+        if self._jsonWebToken != '':
+            request.headers['Authorization'] = 'Bearer ' + self._jsonWebToken
+        else:
+            requests_auth.HTTPBasicAuth(self._username, self._password)(request)
+
+            def setJsonWebToken(response, *args, **kwargs):
```
Comment (woswos) [inline python/mujinwebstackclient/controllerwebclientraw.py:57]: Please follow the function naming rules for python

File: python/mujinwebstackclient/controllerwebclientraw.py [hunk 6]
```diff
@@ -47,7 +83,7 @@ def __init__(self, baseurl, username, password, locale=None, author=None, userAg
         self._session = requests.Session()
 
         # Use basic auth
```
Comment (woswos) [inline python/mujinwebstackclient/controllerwebclientraw.py:85]: Let's update the comment

**Comment by woswos** [comment]
Thanks!!

**Review by woswos — APPROVED** [review]

**Comment by LLipter** [comment]
> Question: does this handle jwt expiry correctly?

@ziyan The token never expires at the moment, so this MR does not include a re-try mechanism

**Pull request merged**